api/lib/block-listeners/polkadot-node-subscription.js

```diff
@@ -56,25 +56,35 @@ class PolkadotNodeSubscription {
     }
 
     // Subscribe to new block headers
-    await this.api.rpc.chain.subscribeNewHeads(async (blockHeader) => {
-      const blockHeight = blockHeader.number.toNumber()
-      if (this.height < blockHeight) {
-        this.height = blockHeight
-        this.newBlockHandler(blockHeight)
-      }
+    const unsubscribe = await this.api.rpc.chain.subscribeNewHeads(
+      async (blockHeader) => {
+        const blockHeight = blockHeader.number.toNumber()
+        if (this.height < blockHeight) {
+          this.height = blockHeight
+          this.newBlockHandler(blockHeight) // do not await as this can take some seconds
+        }
 
-      // refresh the api to prevent memory leaks
-      // if (Date.now() - this.store.polkadotRPCOpened > DISCONNECTION_INTERVAL) {
-      //   console.log(
-      //     'Disconnecting Polkadot for network',
-      //     this.network.id,
-      //     'to avoid memory leaks'
-      //   )
-      //   this.api.disconnect()
-      //   this.api = undefined
-      //   await this.initPolkadotRPC()
-      // }
-    })
+        // refresh the api to prevent memory leaks
+        // right after a new block so we don't miss any block
+        if (
+          Date.now() - this.store.polkadotRPCOpened >
+          DISCONNECTION_INTERVAL
+        ) {
+          console.log(
+            'Disconnecting Polkadot for network',
+            this.network.id,
+            'to avoid memory leaks'
+          )
+          // we keep the active connection alive so queries can finish
+          const oldApi = this.api
+          setTimeout(() => oldApi.disconnect(), 1000 * 60 * 5)
```

Comment:
Then after 5 min it really disconnects right? But how do you know there are no other new queries waiting to be finished in the oldApi?

Reply:
new queries use `this.api` which is the new connection